File: src/Asm.MooBank.Web/MooBankApp/src/models/Budget.ts
export interface Budget {
    year: number;
    incomeLines: BudgetLine[];
    expensesLines: BudgetLine[];
    months: BudgetMonth[];
}

export interface BudgetLine {
    id: string,
    name: string,
    tagId: number,
    notes?: string,
    amount: number,
    month?: number,
    type: BudgetLineType,
}

export type BudgetLineType = "income" | "expenses";

export interface BudgetMonth {
    month: number;
    income: number;
    expenses: number;
    remainder: number;
}

export interface BudgetReportByMonth {
    items: BudgetReportValueMonth[];
}

export interface  BudgetReportValue {
    budgetedAmount: number;
    actual: number;
}

export interface BudgetReportValueMonth extends BudgetReportValue {
    month: number;
}

File: src/Asm.MooBank.Models/ModelExtensions/Budget.cs
﻿namespace Asm.MooBank.Models;

public partial record Budget
{
    public static implicit operator Budget(Domain.Entities.Budget.Budget? budget)
    {
        if (budget == null) return null!;

        var lines = budget.Lines.ToModel();
        List<Models.BudgetMonth> months = new();
        for (int i = 0; i < 12; i++)
        {
            var mask = 1 << i;
            var monthIncome = lines.Where(l => l.Type == BudgetLineType.Income && (l.Month & (1 << i)) != 0).Sum(l => l.Amount);
            var monthExpenses = lines.Where(l => l.Type == BudgetLineType.Expenses && (l.Month & (1 << i)) != 0).Sum(l => l.Amount);

            months.Add(new(i, monthIncome, monthExpenses));
        }

        return new()
        {
            Year = budget.Year,
            IncomeLines = lines.Where(l => l.Type == BudgetLineType.Income).OrderBy(l => l.Name),
            ExpensesLines = lines.Where(l => l.Type == BudgetLineType.Expenses).OrderBy(l => l.Name),
            Months = months,
        };
    }
}

public static class BudgetExtensions
{
    public static Domain.Entities.Budget.Budget ToDomain(this Budget budget) =>
        new(Guid.Empty)
        {
            Year = budget.Year,
        };
}